### ISSUE TITLE
IIS 7.5 Binding Bug

### ISSUE THREAD

**Issue by silverBull3t** [Microsoft/IIS.Administration#153]
Hi,
there is a BUG when fetching a website on IIS 7.5 "/api/webserver/websites/bc-sYFqIYoYTtsibA4Tn9A"
my sites have a specific HTTPS binding.
every site that I try to "GET" that has that Binding, I get a 500 error.

----------------this is the log-------------------------
2017-10-16 19:16:24.077 +03:00 [Error] An unhandled exception has occurred: Either the application has not called WSAStartup, or WSAStartup failed. (Exception from HRESULT: 0x8007276D)
System.Runtime.InteropServices.COMException (0x8007276D): Either the application has not called WSAStartup, or WSAStartup failed. (Exception from HRESULT: 0x8007276D)
   at Microsoft.Web.Administration.Interop.IAppHostProperty.get_Value()
   at Microsoft.Web.Administration.ConfigurationElement.GetPropertyValue(IAppHostProperty property)
   at Microsoft.Web.Administration.ConfigurationElement.GetAttributeValue(String attributeName)
   at Microsoft.Web.Administration.Binding.get_CertificateStoreName()
   at Microsoft.IIS.Administration.WebServer.Sites.SiteHelper.ToJsonModel(Binding binding)
   at Microsoft.IIS.Administration.WebServer.Sites.SiteHelper.ToJsonModel(Site site, Fields fields, Boolean full)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionFilterAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeAsync>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.IIS.Administration.WebServer.Injector.<ProceedNoTransaction>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.IIS.Administration.WebServer.Injector.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.IIS.Administration.HeadTransform.<Invoke>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.IIS.Administration.Startup.<>c.<<Configure>b__4_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.IIS.Administration.Security.Authorization.AuthorizationPolicyMiddleware.<Invoke>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.<Invoke>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware`1.<Invoke>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.IIS.Administration.Cors.CorsExtensions.<>c.<<UseCrossOrigin>b__0_3>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware.<Invoke>d__7.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.IIS.Administration.Cors.CorsExtensions.<>c__DisplayClass0_0.<<UseCrossOrigin>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.IIS.Administration.ErrorHandler.<Invoke>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.IIS.Administration.ErrorHandler.<Invoke>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>d__6.MoveNext()
![ssl2](https://user-images.githubusercontent.com/32847013/31623374-e8f80bb6-b2a7-11e7-9b8f-b8aa690382b9.jpg)
![ssl1](https://user-images.githubusercontent.com/32847013/31623375-e91c6330-b2a7-11e7-9464-29d214347bde.jpg)

**Comment by jimmyca15** [comment]
Okay, have you tried creating https bindings with another certificate to see if the issue was limited to that certificate? I see that the binding is using a wildcard subject, is there anything else notable about it?

**Comment by silverBull3t** [comment]
Actsualy no.
Main thing is that it's a wildcard cert.
If i delete the binding from a site. It works. When adding again it breaks

**Comment by silverBull3t** [comment]
Do you have any suggestion for a temp workaround ?
Maybe a change in a ps1 file ?

**Comment by jimmyca15** [comment]
The PowerShell scripts only effect installation so any modifications will not effect runtime behavior. Can you tell me what certificate stores the wildcard certificate is installed in on the machine?

**Comment by silverBull3t** [comment]
I hope i'm answering correctly, all of them a total of 3 certs are in a store called "My"

**Comment by jimmyca15** [comment]
That sounds right, and it is the correct store. The terms "My" and "Personal" are interchangeable.

**Comment by jimmyca15** [comment]
I got onto a server 2008 R2 machine to look into this. I created a self signed certificate using 
`
New-SelfSignedCertificate -Subject "*.abc.com"
`
and then ensured the certificate was in the personal store (My) with the Server Authentication key usage.

After patching a web site to use this certificate for an HTTPS binding I received a 500 error. the binding was created but the certificate was not used in the binding. I was able to patch the web site to use a certificate that did not use a wildcard subject. Also I was **able to update** the website using **Inetmgr.exe** to use the wildcard certificate.

I did not have an exact reproduction of the issue because I had no trouble retrieving the site through the API. The only problem was with creating the binding to use the wildcard certificate. So there **is an issue**.

My log on the PATCH request
```
2017-10-16 11:17:24.469 -07:00 [Error] An unhandled exception has occurred: A specified logon session does not exist. It may already have been terminated. (Exception from HRESULT: 0x80070520)
System.Runtime.InteropServices.COMException (0x80070520): A specified logon session does not exist. It may already have been terminated. (Exception from HRESULT: 0x80070520)
   at Microsoft.Web.Administration.Interop.IAppHostMethodInstance.Execute()
   at Microsoft.Web.Administration.ConfigurationMethodInstance.Execute()
   at Microsoft.Web.Administration.Binding.AddSslCertificate(Byte[] certificateHash, String certificateStoreName)
   at Microsoft.Web.Administration.BindingManager.BindingTransaction.Commit()
   at Microsoft.Web.Administration.BindingManager.Save()
   at Microsoft.Web.Administration.ServerManager.CommitChanges()
   at Microsoft.IIS.Administration.WebServer.MgmtUnit.Commit()
   at Microsoft.IIS.Administration.WebServer.Sites.SitesController.Patch(String id, Object model)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionFilterAsync>d__28.MoveNext()
```

**Comment by silverBull3t** [comment]
Iis 7.5 and wildcard certs are not the best match.
I had to perform some voodoo to enable SNI on it in order to make it work with a header.
It cant be done in the GUI. I use appcmd to attach the binding.
I'm sure that is the root of the problem or atleast a part of it.
The exception you got is different then my.
Would sending you my binding script help ?

**Comment by jimmyca15** [comment]
Yes, add it if you don't mind.

**Comment by jimmyca15** [comment]
Also did you try restarting iis with _iisreset.exe_ and also maybe restart the _Microsoft IIS Administration_ service?
```
c:\>iisreset

Attempting stop...
Internet services successfully stopped
Attempting start...
Internet services successfully restarted

c:\>net stop "Microsoft IIS Administration"
The Microsoft IIS Administration service is stopping.
The Microsoft IIS Administration service was stopped successfully.

c:\>net start "Microsoft IIS Administration"
The Microsoft IIS Administration service is starting.
The Microsoft IIS Administration service was started successfully.
```

After finding an [article ](https://knowledge.geotrust.com/support/knowledge-base/index?page=content&id=SO22115&actp=search&viewlocale=en_US&searchid=1435162914930) that mentions that IIS 7.x may need certificates to be imported with the private key set as exportable I re-imported the certificate I was using and from that point everything was working correctly. 

I am still trying to reproduce the issue that you were seeing, but from the error message it seems that either IIS or the Administration service entered a bad state.

**Comment by silverBull3t** [comment]
appcmd set site /site.name:"%1" /+bindings.[protocol='https',bindingInformation='xxx.xxx.xxx.xxx:443:%1.yyy.com']

**Comment by silverBull3t** [comment]
and yep.
I restarted IIS using iisreset and also restarted IIS Admin Service. that was one of the first things I tried.

**Comment by silverBull3t** [comment]
any ideas ?
maybe connecting remotely to my system and seeing the bug would help?

**Comment by jimmyca15** [comment]
You mentioned that you get an error when trying to create the binding through the inetmgr.exe UI right? Is it the same error message?

**Comment by jimmyca15** [comment]
I just reproduced the issue by using the appcmd command you sent.
```
appcmd set site /site.name:"Default Web Site" /+bindings.[protocol='https',bindingInformation='127.0.0.1:443:abc.abc.com']
```

**Comment by silverBull3t** [comment]
Nope. Actsualy it's not posoible to set this binding from the UI.
This is why i have the script.

So...after using my binding script you get them same exact error ?
If so this is good news. I hope a patch will soon follow.

**Comment by jimmyca15** [comment]
I noticed that the appcmd script that you provided sets a specific value for the ip address in the binding information attribute. This appears to be what is causing the issue. Do you have multiple ip addresses assigned to the server that you are creating the binding on?

**Comment by jimmyca15** [comment]
If you change your script to listen on all IP the problem should disappear

`
appcmd set site /site.name:"%1" /+bindings.[protocol='https',bindingInformation='*:443:%1.yyy.com']
`

**Comment by silverBull3t** [comment]
I need to try this.
The problem is that i have hundrads of sites allready setup this way. I would need to go over all of them or atleast write a script for that.
Also, i dont know how this would work with multiple certs on other ip addresses. I would need to do some testing to make sure it doesn't break.

Also, there is a big chance that other users will encouter this bug. I believe it would be better to have a fix for this issue instead of going through hundrads of sites in a production server.

What do you think ?

**Comment by jimmyca15** [comment]
I agree that the problem should be addressed. Specifying any IP is a workaround of sorts to make the API operable. 

It looks like the problem is not something that will have a simple work around to accomplish the same level of functionality. I have not yet dived into the code to get the root cause but it could be far down in the stack.

**Comment by jimmyca15** [comment]
@silverBull3t 

We have found out what is causing this issue to occur on IIS 7.5 machines. There will be a fix for this in our next release.

**Comment by silverBull3t** [comment]
great, thank you.
hope the release will come soon.

**Comment by jimmyca15** [comment]
Hello @silverBull3t we just released a new version of the IIS Administration API. The 2.2.0 release fixes this issue. 

Fixed by #168
